File: mobile/ios/PortaliaCore/Sources/PortaliaCore/Networking/APIError.swift
import Foundation

public enum APIError: Error, LocalizedError {
    case network(Error)
    case decoding(Error)
    case server(status: Int, code: String?, message: String)
    case unauthorized
    case unknown

    public var errorDescription: String? {
        switch self {
        case .network:
            return "Não foi possível conectar ao servidor."
        case .decoding:
            return "Resposta inesperada do servidor."
        case .server(_, _, let message):
            return message
        case .unauthorized:
            return "Sessão expirada. Entre novamente."
        case .unknown:
            return "Algo deu errado. Tente novamente."
        }
    }
}

/// Espelha o envelope de erro padrão do backend: `{error: {code, message, correlationId, details}}`.
struct ServerErrorEnvelope: Decodable {
    struct Body: Decodable {
        let code: String?
        let message: String
    }
    let error: Body
}

File: mobile/android/core/src/main/kotlin/com/portalia/core/net/ApiError.kt
package com.portalia.core.net

import kotlinx.serialization.Serializable

sealed class ApiError(message: String) : Exception(message) {
    class Network(cause: Throwable) : ApiError("Não foi possível conectar ao servidor.")
    class Decoding(cause: Throwable) : ApiError("Resposta inesperada do servidor.")
    class Server(val status: Int, val code: String?, val serverMessage: String) : ApiError(serverMessage)
    object Unauthorized : ApiError("Sessão expirada. Entre novamente.")
    object Unknown : ApiError("Algo deu errado. Tente novamente.")
}

@Serializable
data class ServerErrorEnvelope(val error: ServerErrorBody)

@Serializable
data class ServerErrorBody(val code: String? = null, val message: String)
